File: 1405 Longest Happy String/1405.py
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        s = ""
        pq = [(-x, chr(ord('a') + i)) for i, x in enumerate([a, b, c]) if x > 0]
        heapify(pq)

        while pq:
            cnt, x  = heappop(pq)

            if len(s) >= 2 and s[-2] == s[-1] and s[-1] == x:
                if not pq: 
                    break
                else:
                    cnt_2, x_2  = heappop(pq)
                    s += x_2
                    cnt_2 += 1
                    if cnt_2 < 0:
                        heappush(pq, (cnt_2, x_2))
                    heappush(pq, (cnt, x))
            else:
                s += x
                cnt += 1
                if cnt < 0:
                    heappush(pq, (cnt, x))

        return s

File: 1405 Longest Happy String/1405.cpp
#define PIC pair<int, char> 
#define PQ priority_queue<PIC> 

class Solution {
public:
    string longestDiverseString(int a, int b, int c) {
        PQ pq;

        if (a > 0) pq.push({a, 'a'});
        if (b > 0) pq.push({b, 'b'});
        if (c > 0) pq.push({c, 'c'});

        string ans = "";

        while (!pq.empty()){
            int cnt = pq.top().first;
            char x = pq.top().second;
            pq.pop();
            int n = ans.size();
            if (n >= 2 && ans[n-2] == ans[n-1] && ans[n-1] == x){
                if (pq.empty()){
                    break;
                } else {
                    int cnt2 = pq.top().first;
                    char x2 = pq.top().second;
                    pq.pop();
                    ans += x2;
                    cnt2--;
                    if (cnt2 > 0) pq.push({cnt2, x2}); 
                    pq.push({cnt, x}); 
                }
            } else {
                ans += x;
                cnt--;
                if (cnt > 0) pq.push({cnt, x});
            }
        }

        return ans;
    }
};
